File: tests/algorithms/transpiler/FS/matrix/max_area_of_island.fs
// Generated 2025-08-17 13:19 +0700

exception Return
let mutable _nowSeed:int64 = 0L
let mutable _nowSeeded = false
let _initNow () =
    let s = System.Environment.GetEnvironmentVariable("MOCHI_NOW_SEED")
    if System.String.IsNullOrEmpty(s) |> not then
        match System.Int32.TryParse(s) with
        | true, v ->
            _nowSeed <- int64 v
            _nowSeeded <- true
        | _ -> ()
let _now () =
    if _nowSeeded then
        _nowSeed <- (_nowSeed * 1664525L + 1013904223L) % 2147483647L
        int _nowSeed
    else
        int (System.DateTime.UtcNow.Ticks % 2147483647L)

_initNow()
let _dictAdd<'K,'V when 'K : equality> (d:System.Collections.Generic.IDictionary<'K,'V>) (k:'K) (v:'V) =
    d.[k] <- v
    d
let _dictCreate<'K,'V when 'K : equality> (pairs:('K * 'V) list) : System.Collections.Generic.IDictionary<'K,'V> =
    let d = System.Collections.Generic.Dictionary<'K, 'V>()
    for (k, v) in pairs do
        d.[k] <- v
    upcast d
let _idx (arr:'a array) (i:int) : 'a =
    if not (obj.ReferenceEquals(arr, null)) && i >= 0 && i < arr.Length then arr.[i] else Unchecked.defaultof<'a>
let rec _str v =
    match box v with
    | :? float as f -> sprintf "%.10g" f
    | _ ->
        let s = sprintf "%A" v
        s.Replace("[|", "[")
         .Replace("|]", "]")
         .Replace("; ", " ")
         .Replace(";", "")
         .Replace("\"", "")
let __bench_start = _now()
let __mem_start = System.GC.GetTotalMemory(true)
open System.Collections.Generic

let rec encode (row: int) (col: int) =
    let mutable __ret : string = Unchecked.defaultof<string>
    let mutable row = row
    let mutable col = col
    try
        __ret <- ((_str (row)) + ",") + (_str (col))
        raise Return
        __ret
    with
        | Return -> __ret
and is_safe (row: int) (col: int) (rows: int) (cols: int) =
    let mutable __ret : bool = Unchecked.defaultof<bool>
    let mutable row = row
    let mutable col = col
    let mutable rows = rows
    let mutable cols = cols
    try
        __ret <- (((row >= 0) && (row < rows)) && (col >= 0)) && (col < cols)
        raise Return
        __ret
    with
        | Return -> __ret
and has (seen: System.Collections.Generic.IDictionary<string, bool>) (key: string) =
    let mutable __ret : bool = Unchecked.defaultof<bool>
    let mutable seen = seen
    let mutable key = key
    try
        __ret <- seen.ContainsKey(key)
        raise Return
        __ret
    with
        | Return -> __ret
and depth_first_search (row: int) (col: int) (seen: System.Collections.Generic.IDictionary<string, bool>) (mat: int array array) =
    let mutable __ret : int = Unchecked.defaultof<int>
    let mutable row = row
    let mutable col = col
    let mutable seen = seen
    let mutable mat = mat
    try
        let rows: int = Seq.length (mat)
        let cols: int = Seq.length (_idx mat (int 0))
        let key: string = encode (row) (col)
        if ((is_safe (row) (col) (rows) (cols)) && (not (has (seen) (key)))) && ((_idx (_idx mat (int row)) (int col)) = 1) then
            seen <- _dictAdd (seen) (string (key)) (true)
            __ret <- (((1 + (depth_first_search (row + 1) (col) (seen) (mat))) + (depth_first_search (row - 1) (col) (seen) (mat))) + (depth_first_search (row) (col + 1) (seen) (mat))) + (depth_first_search (row) (col - 1) (seen) (mat))
            raise Return
        else
            __ret <- 0
            raise Return
        __ret
    with
        | Return -> __ret
and find_max_area (mat: int array array) =
    let mutable __ret : int = Unchecked.defaultof<int>
    let mutable mat = mat
    try
        let mutable seen: System.Collections.Generic.IDictionary<string, bool> = _dictCreate []
        let rows: int = Seq.length (mat)
        let mutable max_area: int = 0
        let mutable r: int = 0
        while r < rows do
            let line: int array = _idx mat (int r)
            let cols: int = Seq.length (line)
            let mutable c: int = 0
            while c < cols do
                if (_idx line (int c)) = 1 then
                    let key: string = encode (r) (c)
                    if not (seen.ContainsKey(key)) then
                        let area: int = depth_first_search (r) (c) (seen) (mat)
                        if area > max_area then
                            max_area <- area
                c <- c + 1
            r <- r + 1
        __ret <- max_area
        raise Return
        __ret
    with
        | Return -> __ret
let matrix: int array array = [|[|0; 0; 1; 0; 0; 0; 0; 1; 0; 0; 0; 0; 0|]; [|0; 0; 0; 0; 0; 0; 0; 1; 1; 1; 0; 0; 0|]; [|0; 1; 1; 0; 1; 0; 0; 0; 0; 0; 0; 0; 0|]; [|0; 1; 0; 0; 1; 1; 0; 0; 1; 0; 1; 0; 0|]; [|0; 1; 0; 0; 1; 1; 0; 0; 1; 1; 1; 0; 0|]; [|0; 0; 0; 0; 0; 0; 0; 0; 0; 0; 1; 0; 0|]; [|0; 0; 0; 0; 0; 0; 0; 1; 1; 1; 0; 0; 0|]; [|0; 0; 0; 0; 0; 0; 0; 1; 1; 0; 0; 0; 0|]|]
ignore (printfn "%d" (find_max_area (matrix)))
let __bench_end = _now()
let __mem_end = System.GC.GetTotalMemory(true)
printfn "{\n  \"duration_us\": %d,\n  \"memory_bytes\": %d,\n  \"name\": \"main\"\n}" ((__bench_end - __bench_start) / 1000) (__mem_end - __mem_start)

File: tests/algorithms/x/PHP/matrix/max_area_of_island.php
<?php
error_reporting(E_ALL & ~E_DEPRECATED);
ini_set('memory_limit', '-1');
function _str($x) {
    if (is_array($x)) {
        $isList = array_keys($x) === range(0, count($x) - 1);
        if ($isList) {
            $parts = [];
            foreach ($x as $v) { $parts[] = _str($v); }
            return '[' . implode(' ', $parts) . ']';
        }
        $parts = [];
        foreach ($x as $k => $v) { $parts[] = _str($k) . ':' . _str($v); }
        return 'map[' . implode(' ', $parts) . ']';
    }
    if (is_bool($x)) return $x ? 'true' : 'false';
    if ($x === null) return 'null';
    return strval($x);
}
function encode($row, $col) {
  global $matrix;
  return _str($row) . ',' . _str($col);
}
function is_safe($row, $col, $rows, $cols) {
  global $matrix;
  return $row >= 0 && $row < $rows && $col >= 0 && $col < $cols;
}
function has($seen, $key) {
  global $matrix;
  return array_key_exists($key, $seen);
}
function depth_first_search($row, $col, &$seen, $mat) {
  global $matrix;
  $rows = count($mat);
  $cols = count($mat[0]);
  $key = encode($row, $col);
  if (is_safe($row, $col, $rows, $cols) && (!has($seen, $key)) && $mat[$row][$col] == 1) {
  $seen[$key] = true;
  return 1 + depth_first_search($row + 1, $col, $seen, $mat) + depth_first_search($row - 1, $col, $seen, $mat) + depth_first_search($row, $col + 1, $seen, $mat) + depth_first_search($row, $col - 1, $seen, $mat);
} else {
  return 0;
}
}
function find_max_area($mat) {
  global $matrix;
  $seen = [];
  $rows = count($mat);
  $max_area = 0;
  $r = 0;
  while ($r < $rows) {
  $line = $mat[$r];
  $cols = count($line);
  $c = 0;
  while ($c < $cols) {
  if ($line[$c] == 1) {
  $key = encode($r, $c);
  if (!(array_key_exists($key, $seen))) {
  $area = depth_first_search($r, $c, $seen, $mat);
  if ($area > $max_area) {
  $max_area = $area;
};
};
}
  $c = $c + 1;
};
  $r = $r + 1;
};
  return $max_area;
}
$matrix = [[0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0], [0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 0, 0, 0], [0, 1, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0], [0, 1, 0, 0, 1, 1, 0, 0, 1, 0, 1, 0, 0], [0, 1, 0, 0, 1, 1, 0, 0, 1, 1, 1, 0, 0], [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0], [0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 0, 0, 0], [0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0]];
echo rtrim(json_encode(find_max_area($matrix), 1344)), PHP_EOL;
